confluent_kafka/src/confluent_kafka.c

```diff
@@ -336,29 +336,58 @@ static PyObject *Message_timestamp (Message *self, PyObject *ignore) {
 static PyObject *Message_headers (Message *self, PyObject *ignore) {
 #ifdef RD_KAFKA_V_HEADERS
 	if (self->headers) {
-        Py_INCREF(self->headers);
+		Py_INCREF(self->headers);
+		return self->headers;
+	} else if (self->c_headers) {
+		self->headers = c_headers_to_py(self->c_headers);
+		Py_INCREF(self->headers);
 		return self->headers;
-    } else if (self->c_headers) {
-        self->headers = c_headers_to_py(self->c_headers);
-        rd_kafka_headers_destroy(self->c_headers);
-        self->c_headers = NULL;
-        Py_INCREF(self->headers);
-        return self->headers;
 	} else {
 		Py_RETURN_NONE;
-    }
+	}
 #else
 		Py_RETURN_NONE;
 #endif
 }
 
+static PyObject *Message_headers_dict (Message *self, PyObject *ignore) {
+#ifdef RD_KAFKA_V_HEADERS
+	const char *keybuf;
+	const void *valbuf;
+	size_t count, len;
+	PyObject *dict, *value;
+
+	if (!self->c_headers)
+		Py_RETURN_NONE;
+
+	dict = PyDict_New();
+	if (!dict)
+		return NULL;
+	// Parse into dictionary.
+	count = rd_kafka_header_cnt(self->c_headers);
+	for (size_t i = 0; i < count;  i++) {
+		rd_kafka_header_get_all(self->c_headers, i, &keybuf, &valbuf, &len);
```

Comment:
check return value, something like this is sufficient:
`for (size_t i = 0; !rd_kafka_header_get_all(..., i, ...) ;  i++) {`